### PR TITLE
Presentation: ViewModel. UseCase and Repository layer handled by Result 

### PR DESCRIPTION
**Pull request by anncode1**
#### Evidence:
image

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -9,7 +9,7 @@
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
-        <activity android:name=".presentation.main.ui.MainActivity">
+        <activity android:name=".presentation.main.MainActivity">
```
Comment (walmyrcarvalho) [inline app/src/main/AndroidManifest.xml:12]: 👏

File: app/src/main/java/com/example/annmovies/domain/usecase/movie/GetMovies.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+import com.example.annmovies.domain.model.Movie
+import com.example.annmovies.utils.Result
+
+/**
+ * Created by anahi.salgado on 23/05/2020
+ */
+interface GetMovies {
+    fun invoke(): Result<List<Movie>>
+}
```
Comment (walmyrcarvalho) [inline app/src/main/java/com/example/annmovies/domain/usecase/movie/GetMovies.kt:11]: Don't forget to add an empty line at the end of your files, so you don't have issues with git! 👍 

You can setup Android Studio to do this, like this:
image

File: app/src/main/java/com/example/annmovies/domain/usecase/movie/GetMovies.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.example.annmovies.domain.usecase.movie
+
+import com.example.annmovies.domain.model.Movie
+import com.example.annmovies.utils.Result
+
+/**
```
Comment (walmyrcarvalho) [inline app/src/main/java/com/example/annmovies/domain/usecase/movie/GetMovies.kt:6]: I think you can also remove these comments, so you can leave your classes a lot cleaner! 👍

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -1,3 +1,4 @@
 <resources>
     <string name="app_name">AnnMovies</string>
+    <string name="loading">Loading...</string>
```
Comment (walmyrcarvalho) [inline app/src/main/res/values/strings.xml:3]: Awesome! Are you adding i18n as well? Maybe spanish would be nice! :)

File: app/src/main/java/com/example/annmovies/utils/Result.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package com.example.annmovies.utils
+
+/**
+ * Created by anahi.salgado on 23/05/2020
+ */
+sealed class Result<T> {
+    data class Success<T>(val value: T): Result<T>()
+    data class Error<T>(val error: Throwable): Result<T>()
+}
```
Comment (walmyrcarvalho) [inline app/src/main/java/com/example/annmovies/utils/Result.kt:9]: You can use the same tip regarding line at the end of the file here! 👍

File: app/src/main/java/com/example/annmovies/presentation/main/MainViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    fun getCategoryStates(): MutableLiveData<CategoryState> {
+        getCategories()
+        return categoryState
+    }
+
+    private fun getCategories() {
+        categoryState.value = Loading(listOf())
+        val result = categoryUseCase.invoke()
+        when (result) {
+            is Result.Success -> categoryState.postValue(Categories(result.value))
```
Comment (walmyrcarvalho) [inline app/src/main/java/com/example/annmovies/presentation/main/MainViewModel.kt:31]: Really clean, love it! 😊

File: app/src/main/java/com/example/annmovies/presentation/main/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import kotlinx.android.synthetic.main.activity_main.*
+
+class MainActivity : AppCompatActivity() {
+
+    private lateinit var mainViewModel: MainViewModel
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        val categoryRepositoryImpl = CategoryRepositoryImpl()
```
Comment (walmyrcarvalho) [inline app/src/main/java/com/example/annmovies/presentation/main/MainActivity.kt:20]: Suggestion: you can add an empty line about this one to improve code organization, so the variable blocks are not read on the same context! 👍

File: app/src/main/java/com/example/annmovies/domain/model/Category.kt [hunk 8]
```diff
@@ -5,5 +5,5 @@ package com.example.annmovies.domain.model
  */
 data class Category(
     private val id: Int,
-    private val name: String
+    val name: String
 )
```
Comment (walmyrcarvalho) [inline app/src/main/java/com/example/annmovies/domain/model/Category.kt:9]: I'm curious: why did you remove private here? 👀

File: app/src/main/java/com/example/annmovies/domain/repository/movie/MovieRepository.kt [hunk 9]
```diff
@@ -1,13 +1,8 @@
...
-import io.reactivex.rxjava3.core.Single
-
 /**
  * Created by anahi.salgado on 23/03/2020
  */
 interface MovieRepository {
-    fun getPopularMovies(): Single<List<Movie>>
-    fun getTopRatedMovies(): Single<List<Movie>>
-    fun getUpcomingMovies(): Single<List<Movie>>
+
```
Comment (walmyrcarvalho) [inline app/src/main/java/com/example/annmovies/domain/repository/movie/MovieRepository.kt:7]: Suggestion: maybe you can add some `TODO` comments to mark what you're going to implement. What you think? 😊

File: app/src/main/java/com/example/annmovies/domain/usecase/movie/popular/GetPopularMoviesImpl.kt [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+import com.example.annmovies.domain.repository.movie.MovieRepository
+import com.example.annmovies.domain.usecase.movie.GetMovies
+import com.example.annmovies.utils.Result
+
+/**
+ * Created by anahi.salgado on 15/03/2020
+ */
+class GetPopularMoviesImpl(private val movieRepository: MovieRepository): GetMovies {
+    override fun invoke(): Result<List<Movie>> {
+        TODO("Not yet implemented")
```
Comment (walmyrcarvalho) [inline app/src/main/java/com/example/annmovies/domain/usecase/movie/popular/GetPopularMoviesImpl.kt:13]: This! 👏

File: app/src/main/java/com/example/annmovies/domain/usecase/movie/GetMovies.kt (hunk 2 above)
Comment (anncode1) [reply app/src/main/java/com/example/annmovies/domain/usecase/movie/GetMovies.kt:11]: omg! I didn't know it! this is very useful 
Thank you

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (anncode1) [reply app/src/main/res/values/strings.xml:3]: Cool idea I'll do

File: app/src/main/java/com/example/annmovies/presentation/main/MainViewModel.kt (hunk 6 above)
Comment (anncode1) [reply app/src/main/java/com/example/annmovies/presentation/main/MainViewModel.kt:31]: Yay! 🙌

File: app/src/main/java/com/example/annmovies/presentation/main/MainActivity.kt (hunk 7 above)
Comment (anncode1) [reply app/src/main/java/com/example/annmovies/presentation/main/MainActivity.kt:20]: For sure, I'm going to implement Dagger then this code block will be cleaner

File: app/src/main/java/com/example/annmovies/domain/repository/movie/MovieRepository.kt (hunk 9 above)
Comment (anncode1) [reply app/src/main/java/com/example/annmovies/domain/repository/movie/MovieRepository.kt:7]: For sure, thanks

File: app/src/main/java/com/example/annmovies/domain/model/Category.kt (hunk 8 above)
Comment (anncode1) [reply app/src/main/java/com/example/annmovies/domain/model/Category.kt:9]: If I didn't I was no able to print name field, look this: https://github.com/anncode1/AnnMovies/pull/8/files#diff-cd735d23499d24c48d45e36be50fd8bcR37

Do you have any suggestion on how can I do this?

File: app/src/main/java/com/example/annmovies/domain/repository/movie/MovieRepository.kt (hunk 9 above)
Comment (anncode1) [reply app/src/main/java/com/example/annmovies/domain/repository/movie/MovieRepository.kt:7]: For sure, thanks
